src/Microsoft.Health.Fhir.SqlServer/Features/Search/Expressions/Visitors/PartitionEliminationRewriter.cs

```diff
@@ -0,0 +1,186 @@
+            ISearchParameterDefinitionManager.SearchableSearchParameterDefinitionManagerResolver searchParameterDefinitionManagerResolver)
+        {
+            EnsureArg.IsNotNull(model, nameof(model));
+            EnsureArg.IsNotNull(schemaInformation, nameof(schemaInformation));
+            EnsureArg.IsNotNull(searchParameterDefinitionManagerResolver, nameof(searchParameterDefinitionManagerResolver));
+
+            _model = model;
+            _schemaInformation = schemaInformation;
+            _resourceTypeSearchParameter = searchParameterDefinitionManagerResolver.Invoke().GetSearchParameter(KnownResourceTypes.Resource, SearchParameterNames.ResourceType);
+        }
+
+        private SearchParameterExpression AllTypesExpression
+        {
+            get
+            {
+                if (_allTypesExpression != null)
+                {
+                    return _allTypesExpression;
+                }
+
+                var typeExpressions = new Expression[_model.ResourceTypeIdRange.highestId - _model.ResourceTypeIdRange.lowestId + 1];
+
+                for (short i = 0, typeId = _model.ResourceTypeIdRange.lowestId; typeId <= _model.ResourceTypeIdRange.highestId; typeId++, i++)
+                {
+                    typeExpressions[i] = Expression.StringEquals(FieldName.TokenCode, null, _model.GetResourceTypeName(typeId), false);
+                }
+
+                _allTypesExpression = Expression.SearchParameter(
+                    _resourceTypeSearchParameter,
+                    Expression.Or(typeExpressions));
+
+                return _allTypesExpression;
+            }
+        }
+
+        public override Expression VisitSqlRoot(SqlRootExpression expression, object context)
+        {
+            // Look for primary key continuation token (PrimaryKeyParameter) or _type parameters
+
+            int primaryKeyValueIndex = -1;
+            bool hasTypeRestriction = false;
+            for (var i = 0; i < expression.ResourceTableExpressions.Count; i++)
+            {
+                SearchParameterInfo parameter = expression.ResourceTableExpressions[i].Parameter;
+
+                if (ReferenceEquals(parameter, SqlSearchParameters.PrimaryKeyParameter))
+                {
+                    primaryKeyValueIndex = i;
+                }
+                else if (ReferenceEquals(parameter, _resourceTypeSearchParameter))
+                {
+                    hasTypeRestriction = true;
+                }
+            }
+
+            if (primaryKeyValueIndex < 0)
+            {
+                // no continuation token
+
+                if (hasTypeRestriction)
+                {
+                    // This is already constrained to be one or more resource types.
+                    return expression;
+                }
+
+                if (_schemaInformation.Current < SchemaVersionConstants.PartitionedTables)
+                {
+                    return expression;
+                }
```

Comment:
nit: Is there a reason we don't want this check at the beginning of the method?

Reply:
Good point.